controller.py

```diff
@@ -406,12 +407,24 @@ def add_items(self, collection_details):
             )
             messages.append(message)
 
-        for urn, identifier in identifiers_by_urn.items():
-            status = HTTP_OK
-            description = "Already in catalog"
+        # Find the subset of incoming identifiers that are already
+        # in the catalog.
+        already_in_catalog, ignore = self._in_catalog_subset(
+            collection, identifiers_by_urn
+        )
 
-            if identifier not in collection.catalog:
-                collection.catalog_identifier(identifier)
+        # Everything else needs to be added to the catalog.
+        needs_to_be_added = [
+            x for x in identifiers_by_urn.values()
+            if x not in already_in_catalog
```

Comment:
If `already_in_catalog` is a list of ids, I think this clause should be `if x.id not in already_in_catalog`... but it appears to work in the tests?

Reply:
The test didn't catch it because the test checks the messages, and when it came time to generate the messages, the code correctly checked the identifier IDs. This method still gave the right results, but some useless work was done because `catalog_identifiers` was called on all the identifiers, even the ones already in the catalog.